diff --git a/lib/kobject_uevent.c b/lib/kobject_uevent.c
index 8919b43..ec6f774 100644
--- a/lib/kobject_uevent.c
+++ b/lib/kobject_uevent.c
@@ -65,6 +65,18 @@ static const char *kobject_actions[] = {
 	[KOBJ_BYPASS] =         "bypass",
 };
 
+u64 uevent_next_seqnum(void)
+{
+	u64 seq;
+
+	mutex_lock(&uevent_sock_mutex);
+	seq = ++uevent_seqnum;
+	mutex_unlock(&uevent_sock_mutex);
+
+	return seq;
+}
+EXPORT_SYMBOL_GPL(uevent_next_seqnum);
+
 /**
  * kobject_action_type - translate action string to numeric type
  *
