diff --git a/examples/test-launch.c b/examples/test-launch.c
index 03bcd41..b79c1ee 100644
--- a/examples/test-launch.c
+++ b/examples/test-launch.c
@@ -58,7 +58,7 @@ main (int argc, char *argv[])
   /* create a server instance */
   server = gst_rtsp_server_new ();
   g_object_set (server, "service", port, NULL);
-
+  gst_rtsp_server_set_address(server, argv[1]);
   /* get the mount points for this server, every server has a default object
    * that be used to map uri mount points to media factories */
   mounts = gst_rtsp_server_get_mount_points (server);
@@ -68,7 +68,7 @@ main (int argc, char *argv[])
    * any launch line works as long as it contains elements named pay%d. Each
    * element with pay%d names will be a stream */
   factory = gst_rtsp_media_factory_new ();
-  gst_rtsp_media_factory_set_launch (factory, argv[1]);
+  gst_rtsp_media_factory_set_launch (factory, argv[2]);
 
   /* attach the test factory to the /test url */
   gst_rtsp_mount_points_add_factory (mounts, "/test", factory);
@@ -80,7 +80,7 @@ main (int argc, char *argv[])
   gst_rtsp_server_attach (server, NULL);
 
   /* start serving */
-  g_print ("stream ready at rtsp://127.0.0.1:%s/test\n", port);
+  g_print ("stream ready at rtsp://%s:%s/test\n%s\n",argv[1], port,argv[2]);
   g_main_loop_run (loop);
 
   return 0;
