diff --git a/gst/rtsp-server/rtsp-media-factory.c b/gst/rtsp-server/rtsp-media-factory.c
index 087aa8a..c73406e 100644
--- a/gst/rtsp-server/rtsp-media-factory.c
+++ b/gst/rtsp-server/rtsp-media-factory.c
@@ -127,6 +127,22 @@ static GstElement *default_create_pipeline (GstRTSPMediaFactory * factory,
 
 G_DEFINE_TYPE (GstRTSPMediaFactory, gst_rtsp_media_factory, G_TYPE_OBJECT);
 
+static guint signal_watch_id;
+
+static gboolean
+term_handler (gpointer user_data)
+{
+  GstElement *pipeline = (GstElement *) user_data;
+
+  gst_element_set_state(pipeline, GST_STATE_NULL);
+  gst_object_unref(GST_OBJECT (pipeline));
+
+  /* remove signal handler */
+  signal_watch_id = 0;
+  return FALSE;
+}
+
+
 static void
 gst_rtsp_media_factory_class_init (GstRTSPMediaFactoryClass * klass)
 {
@@ -1177,6 +1193,9 @@ default_create_element (GstRTSPMediaFactory * factory, const GstRTSPUrl * url)
   if (element == NULL)
     goto parse_error;
 
+  signal_watch_id =
+       g_unix_signal_add (SIGTERM, (GSourceFunc) term_handler, element );
+
   GST_RTSP_MEDIA_FACTORY_UNLOCK (factory);
 
   if (error != NULL) {
