diff --git a/arch/mips/pthread_arch.h b/arch/mips/pthread_arch.h
index 904a248..aeac4f2 100644
--- a/arch/mips/pthread_arch.h
+++ b/arch/mips/pthread_arch.h
@@ -1,5 +1,9 @@
 static inline struct pthread *__pthread_self()
 {
+#if defined (__m5281) || defined (__m4281)
+	register char *tp __asm__("$3");
+	__asm__ __volatile__ ("mflxc0\t%0, $8" : "=v" (tp));
+#else
 #if __mips_isa_rev < 2
 	register char *tp __asm__("$3");
 	__asm__ __volatile__ (".word 0x7c03e83b" : "=r" (tp) );
@@ -7,6 +11,7 @@ static inline struct pthread *__pthread_self()
 	char *tp;
 	__asm__ __volatile__ ("rdhwr %0, $29" : "=r" (tp) );
 #endif
+#endif
 	return (pthread_t)(tp - 0x7000 - sizeof(struct pthread));
 }
 
