--- a/src/include/memops.h
+++ b/src/include/memops.h
@@ -91,14 +91,19 @@ do {								\
 } while(0)
 
 #endif /* GNU C and x86 */
+extern void* _memcpy(Uint8 * dst, Uint8 * src, unsigned int len);
 
 /* If there are no optimized versions, define the normal versions */
 #ifndef GAL_memcpy
-#define GAL_memcpy(dst, src, len)	memcpy(dst, src, len)
+//#define GAL_memcpy(dst, src, len)	memcpy(dst, src, len)
+#define GAL_memcpy(dst, src, len)	_memcpy(dst, src, len)
 #endif
 
+
+
 #ifndef GAL_memcpy4
-#define GAL_memcpy4(dst, src, len)	memcpy(dst, src, (len) << 2)
+//#define GAL_memcpy4(dst, src, len)	memcpy(dst, src, (len) << 2)
+#define GAL_memcpy4(dst, src, len)	_memcpy(dst, src, (len) << 2)
 #endif
 
 #ifndef GAL_revcpy
--- a/src/newgal/surface.c
+++ b/src/newgal/surface.c
@@ -18,6 +18,14 @@
 #include "memops.h"
 #include "leaks.h"
 
+void* _memcpy(Uint8 * dst, Uint8 * src, unsigned int len)
+{
+    Uint8 * ret=dst;
+    while(len--)
+        *dst++=*src++;
+    return (void*)ret;
+}
+
 /* Public routines */
 /*
  * Create an empty RGB surface of the appropriate depth
