--- a/drivers/net/wireless/realtek/rtkwifiu/rtl8192eu/hal/hal_dm_acs.c
+++ b/drivers/net/wireless/realtek/rtkwifiu/rtl8192eu/hal/hal_dm_acs.c
@@ -105,7 +105,7 @@ void rtw_acs_trigger(_adapter *adapter,
 
 	hal_data->acs.trigger_ch = scan_chan;
 	/*scan_time - ms ,1ms can sample 250 times*/
-	sample_times = scan_time_ms * 250;
+	sample_times = scan_time_ms * 1;
 	phydm_ccx_monitor_trigger(phydm, sample_times);
 
 	#ifdef CONFIG_RTW_ACS_DBG
--- a/drivers/net/wireless/realtek/rtkwifiu/rtl8723bs/hal/hal_dm_acs.c
+++ b/drivers/net/wireless/realtek/rtkwifiu/rtl8723bs/hal/hal_dm_acs.c
@@ -105,7 +105,7 @@ void rtw_acs_trigger(_adapter *adapter,
 
 	hal_data->acs.trigger_ch = scan_chan;
 	/*scan_time - ms ,1ms can sample 250 times*/
-	sample_times = scan_time_ms * 250;
+	sample_times = scan_time_ms * 1;
 	phydm_ccx_monitor_trigger(phydm, sample_times);
 
 	#ifdef CONFIG_RTW_ACS_DBG
--- a/drivers/net/wireless/realtek/rtkwifiu/rtl8723ds/hal/hal_dm_acs.c
+++ b/drivers/net/wireless/realtek/rtkwifiu/rtl8723ds/hal/hal_dm_acs.c
@@ -105,7 +105,7 @@ void rtw_acs_trigger(_adapter *adapter,
 
 	hal_data->acs.trigger_ch = scan_chan;
 	/*scan_time - ms ,1ms can sample 250 times*/
-	sample_times = scan_time_ms * 250;
+	sample_times = scan_time_ms * 1;
 	phydm_ccx_monitor_trigger(phydm, sample_times);
 
 	#ifdef CONFIG_RTW_ACS_DBG
--- a/drivers/net/wireless/realtek/rtkwifiu/rtl8812au/hal/hal_dm_acs.c
+++ b/drivers/net/wireless/realtek/rtkwifiu/rtl8812au/hal/hal_dm_acs.c
@@ -105,7 +105,7 @@ void rtw_acs_trigger(_adapter *adapter,
 
 	hal_data->acs.trigger_ch = scan_chan;
 	/*scan_time - ms ,1ms can sample 250 times*/
-	sample_times = scan_time_ms * 250;
+	sample_times = scan_time_ms * 1;
 	phydm_ccx_monitor_trigger(phydm, sample_times);
 
 	#ifdef CONFIG_RTW_ACS_DBG
--- a/drivers/net/wireless/realtek/rtkwifiu/rtl8821cs/hal/hal_dm_acs.c
+++ b/drivers/net/wireless/realtek/rtkwifiu/rtl8821cs/hal/hal_dm_acs.c
@@ -105,7 +105,7 @@ void rtw_acs_trigger(_adapter *adapter,
 
 	hal_data->acs.trigger_ch = scan_chan;
 	/*scan_time - ms ,1ms can sample 250 times*/
-	sample_times = scan_time_ms * 250;
+	sample_times = scan_time_ms * 1;
 	phydm_ccx_monitor_trigger(phydm, sample_times);
 
 	#ifdef CONFIG_RTW_ACS_DBG
--- a/drivers/net/wireless/realtek/rtkwifiu/rtl8822be/hal/hal_dm_acs.c
+++ b/drivers/net/wireless/realtek/rtkwifiu/rtl8822be/hal/hal_dm_acs.c
@@ -138,7 +138,7 @@ void rtw_acs_trigger(_adapter *adapter,
 
 	hal_data->acs.trigger_ch = scan_chan;
 	/*scan_time - ms ,1ms can sample 250 times*/
-	sample_times = scan_time_ms * 250;
+	sample_times = scan_time_ms * 1;
 	phydm_ccx_monitor_trigger(phydm, sample_times);
 
 	#ifdef CONFIG_RTW_ACS_DBG
--- a/drivers/net/wireless/realtek/rtkwifiu/rtl8822bs/hal/hal_dm_acs.c
+++ b/drivers/net/wireless/realtek/rtkwifiu/rtl8822bs/hal/hal_dm_acs.c
@@ -138,7 +138,7 @@ void rtw_acs_trigger(_adapter *adapter,
 
 	hal_data->acs.trigger_ch = scan_chan;
 	/*scan_time - ms ,1ms can sample 250 times*/
-	sample_times = scan_time_ms * 250;
+	sample_times = scan_time_ms * 1;
 	phydm_ccx_monitor_trigger(phydm, sample_times);
 
 	#ifdef CONFIG_RTW_ACS_DBG
