--- a/drivers/net/wireless/realtek/rtkwifiu/rtl8822bs/core/rtw_mlme_ext.c
+++ b/drivers/net/wireless/realtek/rtkwifiu/rtl8822bs/core/rtw_mlme_ext.c
@@ -565,8 +565,8 @@ static u8 init_channel_set(_adapter *pad
 
 	_rtw_memset(channel_set, 0, sizeof(RT_CHANNEL_INFO) * MAX_CHANNEL_NUM);
 
-	if (IsSupported24G(regsty->wireless_mode) && hal_chk_band_cap(padapter, BAND_CAP_2G))
-		b2_4GBand = _TRUE;
+	//if (IsSupported24G(regsty->wireless_mode) && hal_chk_band_cap(padapter, BAND_CAP_2G))
+	//	b2_4GBand = _TRUE;
 
 	if (is_supported_5g(regsty->wireless_mode) && hal_chk_band_cap(padapter, BAND_CAP_5G))
 		b5GBand = _TRUE;
