--- a/drivers/net/wireless/realtek/rtkwifiu/rtl8821ce/Makefile
+++ b/drivers/net/wireless/realtek/rtkwifiu/rtl8821ce/Makefile
@@ -822,7 +822,7 @@ endif
 
 ########### HAL_RTL8821C #################################
 ifeq ($(CONFIG_RTL8821C), y)
-include $(TopDIR)/rtl8821c.mk
+include $(TopDIR)/rtl8821ce/rtl8821c.mk
 endif
 
 ########### AUTO_CFG  #################################
--- a/drivers/net/wireless/realtek/rtkwifiu/rtl8812au/Makefile
+++ b/drivers/net/wireless/realtek/rtkwifiu/rtl8812au/Makefile
@@ -244,7 +244,7 @@ _BTC_FILES += hal/btc/halbtc8192e1ant.o
 				hal/btc/halbtc8821c2ant.o
 endif
 
-include $(TopDIR)/hal/phydm/phydm.mk
+include $(TopDIR)/rtl8812au/hal/phydm/phydm.mk
 ########### HAL_RTL8188E #################################
 ifeq ($(CONFIG_RTL8188E), y)
 
--- a/drivers/net/wireless/realtek/rtkwifiu/rtl8723bs/Makefile
+++ b/drivers/net/wireless/realtek/rtkwifiu/rtl8723bs/Makefile
@@ -245,7 +245,7 @@ _BTC_FILES += hal/btc/halbtc8192e1ant.o
 				hal/btc/halbtc8821c2ant.o
 endif
 
-include $(TopDIR)/hal/phydm/phydm.mk
+include $(TopDIR)/rtl8723bs/hal/phydm/phydm.mk
 ########### HAL_RTL8188E #################################
 ifeq ($(CONFIG_RTL8188E), y)
 
--- a/drivers/net/wireless/realtek/rtkwifiu/rtl8723ds/Makefile
+++ b/drivers/net/wireless/realtek/rtkwifiu/rtl8723ds/Makefile
@@ -244,7 +244,7 @@ _BTC_FILES += hal/btc/halbtc8192e1ant.o
 				hal/btc/halbtc8821c2ant.o
 endif
 
-include $(TopDIR)/hal/phydm/phydm.mk
+include $(TopDIR)/rtl8723ds/hal/phydm/phydm.mk
 ########### HAL_RTL8188E #################################
 ifeq ($(CONFIG_RTL8188E), y)
 
--- a/drivers/net/wireless/realtek/rtkwifiu/rtl8812bu/Makefile
+++ b/drivers/net/wireless/realtek/rtkwifiu/rtl8812bu/Makefile
@@ -804,7 +804,7 @@ endif
 
 ########### HAL_RTL8822B #################################
 ifeq ($(CONFIG_RTL8822B), y)
-include $(TopDIR)/rtl8822b.mk
+include $(TopDIR)/rtl8812bu/rtl8822b.mk
 endif
 
 ########### HAL_RTL8821C #################################
--- a/drivers/net/wireless/realtek/rtkwifiu/rtl8192eu/Makefile
+++ b/drivers/net/wireless/realtek/rtkwifiu/rtl8192eu/Makefile
@@ -244,7 +244,7 @@ _BTC_FILES += hal/btc/halbtc8192e1ant.o
 				hal/btc/halbtc8821c2ant.o
 endif
 
-include $(TopDIR)/hal/phydm/phydm.mk
+include $(TopDIR)/rtl8192eu/hal/phydm/phydm.mk
 ########### HAL_RTL8188E #################################
 ifeq ($(CONFIG_RTL8188E), y)
 
--- a/drivers/net/wireless/realtek/rtkwifiu/rtl8192ee/Makefile
+++ b/drivers/net/wireless/realtek/rtkwifiu/rtl8192ee/Makefile
@@ -246,7 +246,7 @@ _BTC_FILES += hal/btc/halbtc8192e1ant.o \
 				hal/btc/halbtc8821c2ant.o
 endif
 
-include $(TopDIR)/hal/phydm/phydm.mk
+include $(TopDIR)/rtl8192ee/hal/phydm/phydm.mk
 ########### HAL_RTL8188E #################################
 ifeq ($(CONFIG_RTL8188E), y)
 
