--- a/drivers/net/wireless/realtek/rtkwifiu/rtl8822bs/os_dep/linux/mlme_linux.c
+++ b/drivers/net/wireless/realtek/rtkwifiu/rtl8822bs/os_dep/linux/mlme_linux.c
@@ -72,6 +72,14 @@ void rtw_os_indicate_connect(_adapter *a
 	if (adapter->pid[2] != 0)
 		rtw_signal_process(adapter->pid[2], SIGALRM);
 
+	if(check_fwstate(&adapter->mlmepriv, WIFI_STATION_STATE)) {
+#if (LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 12))
+		kobject_uevent(&adapter->pnetdev->dev.kobj, KOBJ_LINKUP);
+#else
+		kobject_hotplug(&adapter->pnetdev->class_dev.kobj, KOBJ_LINKUP);
+#endif
+	}
+
 #ifdef RTK_DMP_PLATFORM
 	_set_workitem(&adapter->mlmepriv.Linkup_workitem);
 #endif
@@ -164,6 +172,14 @@ void rtw_os_indicate_disconnect(_adapter
 
 	rtw_indicate_wx_disassoc_event(adapter);
 
+	if(check_fwstate(&adapter->mlmepriv, WIFI_STATION_STATE)) {
+#if (LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 12))
+		kobject_uevent(&adapter->pnetdev->dev.kobj, KOBJ_LINKDOWN);
+#else
+		kobject_hotplug(&adapter->pnetdev->class_dev.kobj, KOBJ_LINKDOWN);
+#endif
+	}
+
 #ifdef RTK_DMP_PLATFORM
 	_set_workitem(&adapter->mlmepriv.Linkdown_workitem);
 #endif
--- a/drivers/net/wireless/realtek/rtkwifiu/rtl8192eu/os_dep/linux/mlme_linux.c
+++ b/drivers/net/wireless/realtek/rtkwifiu/rtl8192eu/os_dep/linux/mlme_linux.c
@@ -72,6 +72,14 @@ void rtw_os_indicate_connect(_adapter *a
 	if (adapter->pid[2] != 0)
 		rtw_signal_process(adapter->pid[2], SIGALRM);
 
+	if(check_fwstate(&adapter->mlmepriv, WIFI_STATION_STATE)) {
+#if (LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 12))
+		kobject_uevent(&adapter->pnetdev->dev.kobj, KOBJ_LINKUP);
+#else
+		kobject_hotplug(&adapter->pnetdev->class_dev.kobj, KOBJ_LINKUP);
+#endif
+	}
+
 #ifdef RTK_DMP_PLATFORM
 	_set_workitem(&adapter->mlmepriv.Linkup_workitem);
 #endif
@@ -164,6 +172,14 @@ void rtw_os_indicate_disconnect(_adapter
 
 	rtw_indicate_wx_disassoc_event(adapter);
 
+	if(check_fwstate(&adapter->mlmepriv, WIFI_STATION_STATE)) {
+#if (LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 12))
+		kobject_uevent(&adapter->pnetdev->dev.kobj, KOBJ_LINKDOWN);
+#else
+		kobject_hotplug(&adapter->pnetdev->class_dev.kobj, KOBJ_LINKDOWN);
+#endif
+	}
+
 #ifdef RTK_DMP_PLATFORM
 	_set_workitem(&adapter->mlmepriv.Linkdown_workitem);
 #endif
